src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets

```diff
@@ -67,13 +67,22 @@ Copyright (c) .NET Foundation. All rights reserved.
     <SelfContained>$(PublishSelfContained)</SelfContained>
   </PropertyGroup>
 
+  <!-- Automatically infer the RuntimeIdentifier for properties that require it.
+  SelfContained without a RID is a no-op and semantically hints towards the fact that it can change the behavior of build, publish, and friends.
+  ... So, we infer the RID for SelfContained regardless of the context.
+
+  The other publish properties are specifically labelled Publish* and don't 'NEED' their RID unless we are doing a publish, so the RID inference
+  ... for these properties is limited to publishing only scenarios.
+
+  Finally, library projects and non-executable projects have awkward interactions here so they are excluded.-->
   <PropertyGroup Condition="'$(UseCurrentRuntimeIdentifier)' == ''">
     <UseCurrentRuntimeIdentifier Condition="
                  '$(RuntimeIdentifier)' == '' and
                  '$(_IsExecutable)' == 'true' and '$(IsTestProject)' != 'true' and
                  '$(IsRidAgnostic)' != 'true' and
                  ('$(_IsPublishing)' == 'true' or '$(SelfContained)' == 'true') and
                   (
+                    '$(SelfContained)' == 'true' or
```

Comment:
The `SelfContained` check is partially duplicated, maybe this is a bit simpler:

```xml
  <PropertyGroup Condition="'$(UseCurrentRuntimeIdentifier)' == ''">
    <UseCurrentRuntimeIdentifier Condition="
                 '$(RuntimeIdentifier)' == '' and
                 '$(_IsExecutable)' == 'true' and '$(IsTestProject)' != 'true' and
                 '$(IsRidAgnostic)' != 'true' and
                 ('$(SelfContained)' == 'true' or
                   ('$(_IsPublishing)' == 'true') and
                     (
                        '$(PublishReadyToRun)' == 'true' or
                        '$(PublishSingleFile)' == 'true' or
                        '$(PublishAot)' == 'true'
                      )
                  )">true</UseCurrentRuntimeIdentifier>
  </PropertyGroup>
```

It's still complicated with a bunch of nested clauses though.  Could we set `SelfContained` to true if `_IsPublishing` is true and any of `PublishReadyToRun`, `PublishSingleFile`, or `PublishAot` are true?  If we did that beforehand, then we could remove the whole `_IsPublishing` clause from this condition and it would be a lot simpler and easier to understand.

Reply:
I like that simplification, thanks for suggesting it.

In response to your question where the 3 `Publish*` properties imply `SelfContained`, I think not. based on this quote from @vitek-karas. 

```
PublishAot actually implies self contained, PublishTrimmed also currently implies self contained (but may not in the future),
PublishSingle file doesn’t imply self-contained but it behaves similarly (needs to download RID specific assets and can only publish for a specific RID),
PublishReadToRun is similar to PublishSingleFile in this regard (needs to download RID specific tooling and can only publish for a specific RID)
```